### PR TITLE
Pages 1 part

### PR DESCRIPTION
**Pull request by MaxDaniLOVE**
### Implemented:
- members page;
- member progress page;
- members tasks page;
- create member feature;
- Edit/Detail/Delete members feature;
- cache data to LocalStorage;

### CODE REVIEW

File: src/pages/MembersPage.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  onFormChange = (e) => {
+    const { value, id } = e.target;
+    this.setState(({ registerData }) => ({
+      registerData: inputsParser(value, id, registerData),
+    }));
+  };
+
+  onAddNewMember = async (member) => {
+    await this.db.addNewUser(member);
+    await this.getMembersData();
```
Comment (maximsan) [inline src/pages/MembersPage.js:76]: it's better to await result and return it or set to state

File: src/pages/MembersPage.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    this.setState({
+      registerData: { ...editedUser },
+      isEditMode: true,
+    });
+  };
+
+  onSubmitEditUser = async () => {
+    const { registerData } = this.state;
+    await this.db.editUserData(registerData);
+    await this.getMembersData();
```
Comment (maximsan) [inline src/pages/MembersPage.js:92]: the same is here

File: src/pages/MembersPage.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    this.onModalOpen();
+    this.setState({
+      registerData: { ...editedUser },
+      isDetailMode: true,
+    });
+  };
+
+  onUserDelete = async (userId) => {
+    await this.db.deleteUser(userId);
+    await this.getMembersData();
```
Comment (maximsan) [inline src/pages/MembersPage.js:107]: and here

File: src/pages/MembersPage.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  };
+
+  render() {
+    const { members, isLoaded, showModal, registerData, isEditMode, isDetailMode } = this.state;
+    const btnStyles = { marginBottom: '1rem' };
```
Comment (maximsan) [inline src/pages/MembersPage.js:112]: add space

File: src/services/Firebase.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  apiKey: process.env.REACT_APP_API_KEY,
+  authDomain: `${process.env.REACT_APP_AUTH_DOMAIN}`,
+  databaseURL: `${process.env.REACT_APP_DATABASE_URL}`,
+  projectId: `${process.env.REACT_APP_PROJECT_ID}`,
+  storageBucket: `${process.env.REACT_APP_STORAGE_BUCKET}`,
+  messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
+  appId: process.env.REACT_APP_APP_ID,
+};
+
+firebase.initializeApp(firebaseConfig);
```
Comment (maximsan) [inline src/services/Firebase.js:13]: move to firebase-config file

File: src/services/Firebase.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+export default class Firebase {
+  database = firebase.firestore();
+
+  addNewUser = async (user) => {
+    try {
+      await this.database.collection('users').add(user);
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:22]: specify it

File: src/services/Firebase.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+  getUserData = async (userId) => {
+    let userData;
+    try {
+      userData = await this.database
+        .collection('users')
+        .doc(userId)
+        .get();
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:34]: specify it

File: src/services/Firebase.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    return userData.data();
+  };
+
+  getUsersData = async () => {
+    let usersData;
+    try {
+      usersData = await this.database.collection('users').get();
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:44]: specify it

File: src/services/Firebase.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        .doc(userId)
+        .get();
+    } catch (error) {
+      throw new Error(error);
+    }
+    return userData.data();
+  };
+
+  getUsersData = async () => {
+    let usersData;
```
Comment (maximsan) [inline src/services/Firebase.js:40]: use const

File: src/services/Firebase.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  addNewUser = async (user) => {
+    try {
+      await this.database.collection('users').add(user);
+    } catch (error) {
+      throw new Error(error);
+    }
+  };
+
+  getUserData = async (userId) => {
+    let userData;
```
Comment (maximsan) [inline src/services/Firebase.js:27]: const

File: src/services/Firebase.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+};
+
+firebase.initializeApp(firebaseConfig);
+
+export default class Firebase {
+  database = firebase.firestore();
+
+  addNewUser = async (user) => {
+    try {
+      await this.database.collection('users').add(user);
```
Comment (maximsan) [inline src/services/Firebase.js:20]: it's better to return smth in sucessfull result

File: src/services/Firebase.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    return usersData;
+  };
+
+  getUsersTasks = async (id) => {
+    const allData = [];
+    const data = await this.database
+      .collection('usersTasks')
+      .where('userId', '==', id)
+      .get();
+    data.forEach((task) => {
```
Comment (maximsan) [inline src/services/Firebase.js:55]: use map

File: src/services/Firebase.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    });
+    const tasksInfo = await Promise.all(withAllData);
+    return allData.map((el, idx) => ({ ...el, tasksInfo: tasksInfo[idx] }));
+  };
+
+  getTasks = async (id) => {
+    const taskData = await this.database
+      .collection('tasks')
+      .doc(id)
+      .get();
```
Comment (maximsan) [inline src/services/Firebase.js:71]: add space before return everywhere

File: src/services/Firebase.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+  deleteUser = async (id) => {
+    // TODO add deliting users tasks and progress
+    try {
+      await this.database
+        .collection('users')
+        .doc(id)
+        .delete();
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:109]: specify it

File: src/services/Firebase.js [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  };
+
+  editUserData = async (newData) => {
+    try {
+      await this.database
+        .collection('users')
+        .doc(newData.id)
+        .set(newData, { merge: true });
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:97]: specify it

File: src/services/Firebase.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+  getUsersProgress = async (id) => {
+    let userProgress;
+    try {
+      userProgress = await this.database
+        .collection('usersProgress')
+        .where('userId', '==', id)
+        .get();
+    } catch (error) {
+      throw new Error(error);
```
Comment (maximsan) [inline src/services/Firebase.js:83]: specify it

File: src/utils/cache.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+const addCache = (key, data) => {
+  const stringifyedData = JSON.stringify(data);
+  localStorage.setItem(key, stringifyedData);
```
Comment (maximsan) [inline src/utils/cache.js:3]: do in one line

File: src/utils/cache.js [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+};
+
+const loadCache = (key) => {
+  if (!localStorage[key]) {
+    return null;
+  }
+  const parsedData = JSON.parse(localStorage[key]);
+  return parsedData;
```
Comment (maximsan) [inline src/utils/cache.js:11]: do in one line with ternary operator

File: src/utils/convertDate.js [hunk 19]
```diff
@@ -0,0 +1,11 @@
+const dateToString = (milliseconds) => {
+  const date = new Date(milliseconds).toISOString();
+  return date.substr(0, 10);
+};
+
+const stringToDate = (string) => {
+  const date = Date.parse(string);
+  return date;
```
Comment (maximsan) [inline src/utils/convertDate.js:8]: do in one line

File: src/utils/convertDate.js [hunk 20]
```diff
@@ -0,0 +1,11 @@
+const dateToString = (milliseconds) => {
+  const date = new Date(milliseconds).toISOString();
+  return date.substr(0, 10);
```
Comment (maximsan) [inline src/utils/convertDate.js:3]: do in one line

File: src/utils/inputsParser.js [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+  sex: '',
+  education: '',
+  birthDate: '',
+  universityAverageScore: '',
+  mathScore: '',
+  address: '',
+  mobilePhone: '',
+  skype: '',
+  startDate: '',
+};
```
Comment (maximsan) [inline src/utils/inputsParser.js:27]: move defaultRegisterData to separate file

File: src/pages/MembersTasksPage.js [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    const cachedTasks = loadCache(`${match.params.mid}_tasks`);
+    const cachedName = loadCache(`${match.params.mid}_name`);
+    if (cachedTasks && cachedName) {
+      this.setState({
+        userTasks: cachedTasks,
+        memberName: cachedName,
+        isLoaded: true,
+      });
+    } else {
+      const db = new Firebase();
```
Comment (maximsan) [inline src/pages/MembersTasksPage.js:30]: I guess you can move initialization out of the component

File: src/pages/MembersProgressPage.js [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+    if (cachedProgress && cachedName) {
+      this.setState({
+        progress: cachedProgress,
+        memberName: cachedName,
+        isLoaded: true,
+      });
+    } else {
+      const db = new Firebase();
+      db.getUsersProgress(match.params.mid).then((progress) => {
+        progress.sort((a, b) => (a.trackDate > b.trackDate ? 1 : -1)); // sort from old to new
```
Comment (maximsan) [inline src/pages/MembersProgressPage.js:32]: move to helper method

File: src/pages/MembersProgressPage.js [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    const cachedProgress = loadCache(`${match.params.mid}_progress`);
+    const cachedName = loadCache(`${match.params.mid}_name`);
+    if (cachedProgress && cachedName) {
+      this.setState({
+        progress: cachedProgress,
+        memberName: cachedName,
+        isLoaded: true,
+      });
+    } else {
+      const db = new Firebase();
```
Comment (maximsan) [inline src/pages/MembersProgressPage.js:30]: take it out

File: src/UI/Modal/Modal.js [hunk 25]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:2]: import just createPortal

File: src/UI/Modal/Modal.js [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+            <span className='btn-inner'>Back to grid</span>
+          </Button>
+        </div>
+      </form>
+    </div>
+  );
+  return ReactDOM.createPortal(content, document.getElementById('modal'));
+};
+
+const Modal = (props) => {
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:35]: if you need this component move to separated file

File: src/UI/Modal/Modal.js [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+            </Button>
+          )}
+          <Button onClick={onModalClose} customStyles={{ width: '10rem' }}>
+            <span className='btn-inner'>Back to grid</span>
+          </Button>
+        </div>
+      </form>
+    </div>
+  );
+  return ReactDOM.createPortal(content, document.getElementById('modal'));
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:32]: move document.getElementById('modal') to const out of the component

File: src/UI/Modal/Modal.js [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+import Button from '../Button';
+
+import './Modal.scss';
+
+const ModalOverlay = ({ children, onModalClose, onSubmit, isDetailMode }) => {
+  const content = (
+    <div className='modal-backdrop'>
+      <form className='modal' onSubmit={() => console.log('submit')}>
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:11]: create a method

File: src/UI/Modal/Modal.js [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+        <div className='modal__content'>{children}</div>
+        <div className='modal__footer'>
+          {isDetailMode ? null : (
+            <Button
+              onClick={() => {
+                onSubmit();
+                onModalClose();
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:18]: create  method

File: src/UI/Modal/Modal.js [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+              }}
+              customClass='btn-success'
+            >
+              <p className='btn-inner'>Save</p>
+            </Button>
+          )}
+          <Button onClick={onModalClose} customStyles={{ width: '10rem' }}>
+            <span className='btn-inner'>Back to grid</span>
+          </Button>
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:27]: create  Go Back button

File: src/UI/Modal/Modal.js [hunk 31]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+import Button from '../Button';
+
+import './Modal.scss';
+
+const ModalOverlay = ({ children, onModalClose, onSubmit, isDetailMode }) => {
```
Comment (maximsan) [inline src/UI/Modal/Modal.js:8]: We use DOM here. I believe you need a life cycle here to mount and unmount component on root element, so it's better to re-do it in statefull component. Read it here https://ru.reactjs.org/docs/portals.html

File: src/utils/inputsParser.js [hunk 32]
```diff
@@ -0,0 +1,29 @@
+import { stringToDate } from './convertDate';
+
+const inputsParser = (value, id, data) => {
```
Comment (maximsan) [inline src/utils/inputsParser.js:3]: it's parse only concrete inputs. Rename it or make more generic.

File: src/components/MembersDataModal/MembersDataModal.js [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+    email,
+    lastName,
+    mathScore,
+    mobilePhone,
+    name,
+    sex,
+    skype,
+    startDate,
+    universityAverageScore,
+  } = registerData;
```
[the remark this answers is not in the document]
Comment (MaxDaniLOVE) [reply src/components/MembersDataModal/MembersDataModal.js:19]: this data will be removed at all

File: src/pages/MembersPage.js [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+      });
+    } else {
+      this.getMembersData();
+    }
+  }
+
+  getMembersData = async () => {
+    this.db.getUsersData().then((data) => {
+      const newMembers = [];
+      data.forEach((doc) => {
```
[the remark this answers is not in the document]
Comment (MaxDaniLOVE) [reply src/pages/MembersPage.js:41]: can't use map 'cause firebase returns object not array

File: src/services/Firebase.js (hunk 12 above)
Comment (MaxDaniLOVE) [reply src/services/Firebase.js:55]: can't use map. same as MembersPage.js

File: src/utils/inputsParser.js (hunk 32 above)
Comment (MaxDaniLOVE) [reply src/utils/inputsParser.js:3]: this function will be overwritten

**Pull request closed without merging**